### ISSUE TITLE
Running in Production

### ISSUE THREAD

**Issue by raywan** [dferber90/webapp#4]
After running `npm run build` to build for production, would I still start the app using `npm start` to run `dev.js` i.e use the webpack-dev-server? Sorry I'm a little bit confused.

**Comment by dferber90** [comment]
Sorry for confusing you. This is not documented yet. I will try to explain now. Feel free to ask me to explain more in case it is not clear.

Please also note that this starter-pack was the result of my two-month dive into React, Redux and Webpack. It is my attempt to combine all of these technologies, but I am not quite there yet. I published it now so other people building redux examples can borrow ideas. It is not suitable to be the base of any serious application - at least not yet. Okay, let's answer your questions now :)

> would I still start the app using npm start to run dev.js i.e use the webpack-dev-server?

`webpack-dev-server` should never be used in production.

## how the build process works

When running `npm run build`, `npm run clean` will run first. This will delete everything in `build/` folder so we start with a clean slate. Next, the actual build process will start. This sets
`PRODUCTION=true` and then runs `webpack -p --config webpack/webpack.config.js`. As `webpack/webpack.config.js` exports an array, there will actually be two build processes invoked. One for 
the first exported configuration (client), and one for the second configuration (server).

Webpack will generate files into `build/` folder for both processes. One builds everything the server needs (`backend.js`), while the other process builds every file the client will need later (stuff in `build/public/`), like chunks for gradual loading. So, the build process will generate `build/backend.js` and a bunch of stuff in `build/public/`. You can move all files form `build/` to some other folder and run `node backend.js` in it and it should work. Therefore I will omit the `build` folder in the following text.

`backend.js` is the complete logic of the server. You can run it with `node backend.js`. Now, you can visit `localhost:3000` and verify that your app works. You will see the web app you would normally see. Its HTML includes `<script src=/assets/commonsChunk.js></script>` and other scripts. So the server must serve `commonsChunk.js`. `commonsChunk.js` is generated into `public/assets/commonsChunk.js` along with all the other files necessary for the client. The server (`backend.js`) is configured to serve everything in `public` folder as static content. So `public/assets/commonsChunk.js` can be accessed with `<script src=/assets/commonsChunk.js></script>` [1].

## what's the deal with the `records` folder

You may have noticed the folder `build/records`. This file contains information webpack uses while rebuilding the app during development. You do not need these files when serving in production. The `build`-folder is probably the wrong place for them, but I put them there for now.

## summary

To sum it up, you would need to transfer all files in the `build` directory except the `records`-folder onto your server. Then you need to run `node backend.js` and that's it.

---

[1]: There is a bug right now which causes static files to not be served, because an error is thrown (`Error: initialState was falsy for /assets/commonsChunk.js`) when accessing them. This happens in `generatePage.js:30`. This means it is not possible to run the files in production right now.

**Comment by raywan** [comment]
Thank you so much for the thorough answer!

**Comment by dferber90** [comment]
I've investigated the above mentioned bug. Turns out this is not a bug at all. I was running `node build/backend.js`. This caused the directories to be messed up, because `app.use(express.static('public'))` in `app/server/server.js` looks for the `public` directory relative to the current directory in the command line.

The correct way to run in production is `node backend.js`. You have to be in the folder where `backend.js` is located, otherwise the paths will be messed up.

Leaving this here in case it trips up anybody else.

**Comment by dferber90** [comment]
> You can move all files form build/ to some other folder and run `node backend.js` in it and it should work.

Turns out this is not true, because the `node_modules` are not bundled with the backend.
I'm looking into the build process.

Possible workaround for now is to just copy `node_modules` to the new destination. Downside is that this will also have `devDependencies` installed.
